lnwallet/chanvalidate/validate.go

```diff
@@ -0,0 +1,199 @@
+	Cid lnwire.ShortChannelID
+}
+
+// Locate attempts to locate the funding output within the passed funding
+// transaction.
+//
+// NOTE: Part of the ChanLocator interface.
+func (s *ShortChanIDChanLocator) Locate(fundingTx *wire.MsgTx) (
+	*wire.TxOut, *wire.OutPoint, error) {
+
+	// If the expected index is greater than the amount of output in the
+	// transaction, then we'll reject this channel as it's invalid.
+	outputIndex := s.Cid.TxPosition
+	if int(outputIndex) >= len(fundingTx.TxOut) {
+		return nil, nil, ErrInvalidOutPoint
+	}
+
+	chanPoint := wire.OutPoint{
+		Hash:  fundingTx.TxHash(),
+		Index: uint32(outputIndex),
+	}
+
+	return fundingTx.TxOut[outputIndex], &chanPoint, nil
+}
+
+// CommitmentContext is optional validation context that can be passed into the
+// main Validate for self-owned channel. The information in this context allows
+// us to fully verify out initial commitment spend based on the on-chain state
+// of the funding output.
+type CommitmentContext struct {
+	// Value is the known size of the channel.
+	Value btcutil.Amount
+
+	// FullySignedCommitTx is the fully signed commitment transaction. This
+	// should include a valid witness.
+	FullySignedCommitTx *wire.MsgTx
+}
+
+// Context is the main validation contxet. For a given channel, all fields but
+// the optional CommitCtx should be populated based on existing
+// known-to-be-valid parameters.
+type Context struct {
+	// Locator is a concrete implementation of the ChanLocator interface.
+	Locator ChanLocator
+
+	// MultiSigPkScript is the fully serialized witness script of the
+	// multi-sig output. This is the final witness program that should be
+	// found in the funding output.
+	MultiSigPkScript []byte
+
+	// FundingTx is channel funding transaction as found confirmed in the
+	// chain.
+	FundingTx *wire.MsgTx
+
+	// CommitCtx is an optional additional set of validation context
+	// required to validate a self-owned channel. If present, then fully
+	// Script VM validation will be performed.
+	CommitCtx *CommitmentContext
+}
+
+// Validate given the specified context, this function validates that the
+// alleged channel is well formed, and spendable (if the optional CommitCtx is
+// specified).  If this method returns an error, then the alleged channel is
+// invalid and should be abandoned immediately.
+func Validate(ctx *Context) (*wire.OutPoint, error) {
+	// First, we'll attempt to locate the target outpoint in the funding
+	// transaction. If this returns an error, then we know that the
+	// outpoint doesn't actually exist, so we'll exit early.
+	fundingOutput, chanPoint, err := ctx.Locator.Locate(
+		ctx.FundingTx,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// The scripts should match up exactly, otherwise the channel is
+	// invalid.
+	fundingScript := fundingOutput.PkScript
+	if !bytes.Equal(ctx.MultiSigPkScript, fundingScript) {
+		return nil, ErrWrongPkScript
+	}
+
+	// If there's no commitment context, then we're done here as this is a
```

Comment:
Rather than having an optional commitment context, it would make sense to expose two methods:
`ValidateFundingOutput` and `ValidateCommitment`, where the latter would be the same as this current method, while the first would only consist of the parts up to the optional context (`ValidateCommitment` could just call `ValidateFundingOutput` for its first part, or the caller could just call each in succession).

That would make it less confusing to see that the two uses of this method is doing different validation, and avoids the awkward use of optional fields.

Reply:
I prefer a single method as this means there's a single code path in either scenario where we can focus all tests/review. I don't see how having the context is confusing: it simply gates two extra validation steps. If we want a "proper" optional field, it could very well use a set of functional arguments.

Reply:
I agree that in either case only one function should be called, but I think we can achieve that and still have one validation path. For example, we can expose two distinct methods for each case, that simply wrap the existing `Validate` method and populate the context:

```golang
func ValidateRemote(lnwire.ShortChannelID, []byte, *wire.MsgTx) (*wire.OutPoint, error) {
    return Validate(...)
} 

func ValidateLocal(wire.OutPoint, []byte, *wire.MsgTx, btcutil.Amount, *wire.MsgTx) error {
    _, err := Validate(...)
    return err
} 
```
Would be less verbose, less prone to errors, and more readable at the call site imo. From purely an API/library perspective (if others go on to use this package), I think it's be preferable to avoid the optional field and enforce its presence/absence via a function signature. That being said, we only call this in two locations (+1 for tests) so it wouldn't be saving ourselves very much.

Reply:
Yeah there're two call sites right now, only the wallet call site requires more set up as we need to obtain a signed commitment transaction. I'm not sure the local/remote function really achieves much since the main `Validate` method would still be there. We more or less already have `ValidateLocal` as the funding manager calls `wallet.ValidateChannel` and passes in information similar to your draft function signature above. Also just to put this in perspective, without comments, we're talking about literally 30 lines of code here (the core validate method). We can make a remote validation method for the router, but then that's just more code, making the diff bigger than it needs to be.

Reply:
I agree in this case it's not that big a deal, since we only call this in two places, and the validation logic is fairly simple. But as a general principle I prefer having the fields clearly enforced using a method signature instead of optional fields on a struct, since they tend to be forgotten about and harder to reason about as the struct grows larger.